### PR TITLE
Azure Kubernetes Service sample and helper methods

### PR DESCRIPTION
**Pull request by TsuyoshiUshio**
Based on the conversation of the issue. https://github.com/gruntwork-io/terratest/issues/89
I create a sample and helper method for supporting Azure Kubernetes Service usage. 

This PR provide two helper methods. 

1. Fetch the Kubernetes Service metadata
1. Wait for the allocation of Public IP Address on Kubernetes Service

In case of Azure, Node Port is not good option for external access outside of the kubernetes cluster, we use ingress or LoadBalancer type. For both case, we need to wait until Public IPAddress is assigned to the service. It takes some time. That is why, the sample for kubernetes doesn't work for Azure users. So I'd like to contribute this to unblock it to Azure users.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -156,6 +156,8 @@ such as:
 1.  [Terraform Packer Example](/examples/terraform-packer-example): A more complicated example that shows how to use
     Packer to build an AMI with a web server installed and deploy that AMI in AWS using Terraform.
 1.  [Terraform GCP Example](/examples/terraform-gcp-example): A simple Terraform configuration that creates a GCP Compute Instance and Storage Bucket.
+1.  [Terraform Azure Example](/examples/terraform-azure-example): A simple Terraform configuration that creates an Azure VM.
+1.  [Terraform Azure AKS Example](/examples/terraform-azure-aks-example): A terraform configuration that create Azure Kubernetes Service Cluster
```
Comment (yorinasub17) [inline README.md]:
```suggestion
1.  [Terraform Azure AKS Example](/examples/terraform-azure-aks-example): A terraform configuration that creates an Azure Kubernetes Service (AKS) Cluster.
```

File: examples/terraform-azure-aks-example/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# Terraform Azure AKS Example
+
+This folder contains a simple Terraform moudle that deploys a simple AKS cluster in [Azure](https://azure.microsoft.com/) to demonstrate how you can use Terratest to write automated tests for your Azure Terraform code.
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/README.md]:
```suggestion
This folder contains a Terraform module that deploys a basic AKS cluster in [Azure](https://azure.microsoft.com/) to demonstrate how you can use Terratest to write automated tests for your Azure Terraform code. 
```

We try to avoid the word "simple" because it can be intimidating if you struggle to understand it, and even the most basic examples can be complex to understand if you've never used Azure.

File: examples/terraform-azure-aks-example/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# Terraform Azure AKS Example
+
+This folder contains a simple Terraform moudle that deploys a simple AKS cluster in [Azure](https://azure.microsoft.com/) to demonstrate how you can use Terratest to write automated tests for your Azure Terraform code. 
+
+This module deploies [Azure Kubenetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service/), then deploies nginx by a kubernetes yaml file with Public IPAddress with a service.
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/README.md]:
```suggestion
This module deploys [Azure Kubenetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service/), then deploys nginx by a kubernetes yaml file with a Public IP Address using the `Service` resource.
```

File: examples/terraform-azure-aks-example/README.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+**WARNING**: This module and the automated tests for it deploy real resources into your Azure account which can cost you money. 
+
+## Running this module manually
+
+1. Sign up for [Azure](https://azure.microsoft.com/).
+1. Install [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
+1. Install [Terraform](https://www.terraform.io/) and make sure it's on your `PATH`.
+1. Install [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/) and make sure it's on your `PATH`.
+1. Login Azure with `az login` or `az login --use-device` then configure
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/README.md]:
```suggestion
1. Login to Azure on the CLI with `az login` or `az login --use-device`, and then configure the CLI.
```

File: examples/terraform-azure-aks-example/README.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+1. Create [Service Principal](https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest) then set the value to the environment variables. 
+1. Run `terraform init`
+1. Run `terraform apply`
+1. Apply nginx-deployment.yml
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/README.md]:
```suggestion
1. Apply `nginx-deployment.yml`
```

File: examples/terraform-azure-aks-example/README.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+1. Watch the service until Public IPAddress is assigned.
+1. Send http request to the Public IPAddrees, make sure it returns 200.
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/README.md]:
```suggestion
1. Send http request to the Public IPAddress, make sure it returns 200.
```

File: examples/terraform-azure-aks-example/README.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+1. When you're done, run `terraform destroy`.
+
+### Example
+
+```bash
+$ az login 
+$ export ARM_SUBSCRIPTION_ID={YOUR_SUBSCRIPTION_ID} 
+$ az ad sp create-for-rbac
+$ export TF_VAR_client_id={YOUR_SERVCIE_PRINIPAL_APP_ID}
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/README.md]:
```suggestion
$ export TF_VAR_client_id={YOUR_SERVICE_PRINCIPAL_APP_ID}
```

File: examples/terraform-azure-aks-example/README.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+// Open browser and access the Nginx Service IPAddress
+$ terraform destroy
+```
+
+## Running automated tests against this module
+1. Sign up for [Azure](https://azure.microsoft.com/).
+1. Install [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
+1. Install [Terraform](https://www.terraform.io/) and make sure it's on your `PATH`.
+1. Login Azure with `az login` or `az login --use-device` then configure
+1. Create [Service Principal](https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest) then set the value to the environment variables.
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/README.md]: NIT: Maybe extract these 4 steps out as a prerequisite step, labeled "Prerequisite: Setup Azure CLI access"? Then you can link to the same steps between running example and running test.

File: examples/terraform-azure-aks-example/README.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+1. Install [Golang](https://golang.org/) version `1.13+` required. 
+1. `cd test`
+1. `go test -v -timeout 30m -run TestTerraformAzureAKSExample`
+
+
+### Example
+
+```bash
+$ export ARM_SUBSCRIPTION_ID={YOUR_SUBSCRIPTION_ID} 
+$ export TF_VAR_client_id={YOUR_SERVCIE_PRINIPAL_APP_ID}
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/README.md]:
```suggestion
$ export TF_VAR_client_id={YOUR_SERVICE_PRINCIPAL_APP_ID}
```

File: examples/terraform-azure-aks-example/main.tf [hunk 10]
```diff
@@ -0,0 +1,55 @@
+# ---------------------------------------------------------------------------------------------------------------------
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/main.tf]: NIT: I know many of our examples in this repo don't do this, but for this module it might help to have a preamble with a high level description of what might be deployed. This is because there are more than 2 resources here, so it helps to give some context on the end result. See https://github.com/gruntwork-io/terraform-aws-influx/blob/master/examples/tick-multi-cluster/main.tf#L1 for example (this is one of our public modules that we maintain at Gruntwork).

File: examples/terraform-azure-aks-example/main.tf [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    }
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+# CREATE KUBECONFIG FILE
+# ---------------------------------------------------------------------------------------------------------------------
+
+resource "local_file" "kubeconfig" {
+    content     = azurerm_kubernetes_cluster.k8s.kube_config_raw
+    filename = "kubeconfig"
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/main.tf]: We like to keep our code formatted with `terraform fmt`

File: examples/terraform-azure-aks-example/main.tf (hunk 10 above)
Comment (yorinasub17) [reply examples/terraform-azure-aks-example/main.tf]: NIT: We use two space tabbing for our tf files. It would be good to be consistent.

File: examples/terraform-azure-aks-example/provider.tf [hunk 12]
```diff
@@ -0,0 +1,3 @@
+provider "azurerm" {
+    version = "1.40.0"
+}
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/provider.tf]: NIT: Can you put this in the `main.tf` file? We usually keep small files together to avoid file sprawl. Since the provider is always a single block, we don't typically split it out.

File: examples/terraform-azure-aks-example/variables.tf [hunk 13]
```diff
@@ -0,0 +1,38 @@
+
+
```
Comment (yorinasub17) [inline examples/terraform-azure-aks-example/variables.tf]: NIT: We like to avoid unnecessary whitespace.

Comment (yorinasub17) [reply examples/terraform-azure-aks-example/variables.tf]: Also, we usually provide a description for all variables so that it is clear what should be put in here. E.g I do not know at a glance if `client_secret` refers to secret key for k8s, or for Azure.

File: modules/azure/aks.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2019-11-01/containerservice"
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/gruntwork-io/terratest/modules/retry"
+)
+
+// GetManagedClustersClient is a helper function that will setup an Azure ManagedClusters client on your behalf
+func GetManagedClustersClient(subscriptionID string) (*containerservice.ManagedClustersClient, error) {
```
Comment (yorinasub17) [inline modules/azure/aks.go]: Terratest convention is to use `E` to denote functions that return an `error`. Functions without `E` should use `require` to fail the test if there is an error. This applies to all functions in this file.

File: modules/azure/aks.go [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	managedServicesClient.Authorizer = *authorizer
+
+	return &managedServicesClient, nil
+}
+
+// GetManagedCluster will return ContainerService
```
Comment (yorinasub17) [inline modules/azure/aks.go]: This looks like it is returning a ManagedCluster object, and not ContainerService?

File: modules/azure/aks.go [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+	}
+	managedCluster, err := client.Get(context.Background(), resourceGroupName, clusterName)
+	if err != nil {
+		return nil, err
+	}
+	return &managedCluster, nil
+}
+
+// WaitUntilServiceExternalIPsAvailable is waiting for allocation of External IP Address
+func WaitUntilServiceExternalIPsAvailable(t *testing.T, options *k8s.KubectlOptions, serviceName string, retries int, sleepBetweenRetries time.Duration) {
```
Comment (yorinasub17) [inline modules/azure/aks.go]: This already exists in `k8s`: https://github.com/gruntwork-io/terratest/blob/master/modules/k8s/service.go#L58

File: test/terraform_azure_aks_example_test.go [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+	defer terraform.Destroy(t, terraformOptions)
+
+	// This will run `terraform init` and `terraform apply` and fail the test if there are any errors
+	terraform.InitAndApply(t, terraformOptions)
+
+	// Look up the cluster node count
+	var ActualCount int32 = 0
+	cluster, err := azure.GetManagedCluster(t, expectedResourceGroupName, expectedClusterName, "")
+	assert.Nil(t, err)
+	ActualCount = *(*cluster.ManagedClusterProperties.AgentPoolProfiles)[0].Count
```
Comment (yorinasub17) [inline test/terraform_azure_aks_example_test.go]: We can probably simplify this as:
 
```suggestion
	actualCount := *(*cluster.ManagedClusterProperties.AgentPoolProfiles)[0].Count
```
and remove the `ActualCount` declaration above.

File: test/terraform_azure_aks_example_test.go [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+	// At the end of the test, run `terraform destroy` to clean up any resources that were created
+	defer terraform.Destroy(t, terraformOptions)
+
+	// This will run `terraform init` and `terraform apply` and fail the test if there are any errors
+	terraform.InitAndApply(t, terraformOptions)
+
+	// Look up the cluster node count
+	var ActualCount int32 = 0
+	cluster, err := azure.GetManagedCluster(t, expectedResourceGroupName, expectedClusterName, "")
+	assert.Nil(t, err)
```
Comment (yorinasub17) [inline test/terraform_azure_aks_example_test.go]:
```suggestion
	require.NoError(t, err)
```

File: modules/azure/aks.go (hunk 16 above)
Comment (TsuyoshiUshio) [reply modules/azure/aks.go]: @yorinasub17 It doesn't work for Azure. In case of Azure, we need to use LoadBalancer or ingress type to accept external access. For the Load Balancer with Azure, the method go to the next line if it creates a service, however, we need to wait, until the service is actually get an IP Address. That is why, I create this method. The next question is should I keep it on Azure or should I move it to k8s namespace? If it is general, k8s might be useful for users, however, if it happens only on Azure, I'm wondering where is a good place to put this method.

Comment (yorinasub17) [reply modules/azure/aks.go]: Hmm that seems like a bug if you are seeing this behavior, because it should be doing exactly that: wait for IP address

Can you see what is wrong with https://github.com/gruntwork-io/terratest/blob/master/modules/k8s/service.go#L81?

File: examples/terraform-azure-aks-example/main.tf (hunk 10 above)
Comment (TsuyoshiUshio) [reply examples/terraform-azure-aks-example/main.tf]: Done.  Please review it.

File: examples/terraform-azure-aks-example/main.tf (hunk 11 above)
Comment (TsuyoshiUshio) [reply examples/terraform-azure-aks-example/main.tf]: Done

File: examples/terraform-azure-aks-example/provider.tf (hunk 12 above)
Comment (TsuyoshiUshio) [reply examples/terraform-azure-aks-example/provider.tf]: Done

File: examples/terraform-azure-aks-example/variables.tf (hunk 13 above)
Comment (TsuyoshiUshio) [reply examples/terraform-azure-aks-example/variables.tf]: Done

File: modules/azure/aks.go (hunk 15 above)
Comment (TsuyoshiUshio) [reply modules/azure/aks.go]: Fixed. Done.

File: test/terraform_azure_aks_example_test.go (hunk 17 above)
Comment (TsuyoshiUshio) [reply test/terraform_azure_aks_example_test.go]: Done

File: modules/azure/aks.go (hunk 14 above)
Comment (TsuyoshiUshio) [reply modules/azure/aks.go]: Done

File: modules/azure/aks.go (hunk 16 above)
Comment (TsuyoshiUshio) [reply modules/azure/aks.go]: @yorinasub17  I read the code, it looks exactly the same as mine. 
I suspect that the code is the same, so I remove mine, and point k8s namespace one. 
Maybe the problem is, Even after the IPAddress is allocated, it looks still take time to fetch it.
My test execute polling after the IPAddress readiness, so it might hide the issue. 
 
I'm testing my method ,then I'll replace the method to k8s's and replace it then observe the result, if it looks the same, I'm remove my method.

Comment (TsuyoshiUshio) [reply modules/azure/aks.go]: Hi @yorinasub17 
I figure out what is the problem.  It is not about the `k8s.WaitUntilServiceAvailable()` It works well, however, the problem is, `k8s.GetServiceEndpoint()` The endpoint returns an empty string in case of LoadBalancer type of Azure. Since the method reference Hostname. 

image

The GetServiceEndpointE's documentation said that it return hostname, the behavior works correctly.  

I'm gonna remove WaitFor... method and use `k8s`  and try not to use `GetServiceEndpoint` for Azure.  I'll use 

```go 
	endpoint := service.Status.LoadBalancer.Ingress[0].IP
```
  instead.

File: examples/terraform-azure-aks-example/README.md (hunk 8 above)
Comment (TsuyoshiUshio) [reply examples/terraform-azure-aks-example/README.md]: Done

**Comment by TsuyoshiUshio** [comment]
Hi @yorinasub17  I've done the all fix for your requests. Thank you for your kind review!

File: modules/azure/aks.go (hunk 16 above)
Comment (TsuyoshiUshio) [reply modules/azure/aks.go]: Absolutely, I've done it. Please make sure run CI test for Kubernetes.

**Comment by TsuyoshiUshio** [comment]
Hi @yorinasub17  I've done your request!

**Comment by TsuyoshiUshio** [comment]
@yorinasub17 I remove the merge conflict.

**Comment by yorinasub17** [comment]
@TsuyoshiUshio Thanks! I think code wise this now looks good, but the main sticking point for us right now is the prospect of running automated tests to ensure we can maintain the code going forward. Will post an update in https://github.com/gruntwork-io/terratest/issues/89.

**Comment by TsuyoshiUshio** [comment]
Thank you for the kind review @yorinasub17 , I put the comment on #89

**Comment by yorinasub17** [comment]
We just merged in our tf12 update fix which stabilizes the build. Can you rebase this branch on master to pull in those changes? I can kick off the build once that is done!

**Comment by TsuyoshiUshio** [comment]
Hi @yorinasub17 
Thank you!
I've done the rebase. Please kick off the  build!

**Comment by TsuyoshiUshio** [comment]
Hi, @yorinasub17. I have a look at the root cause of the CI failure. 
It seems Packer related testing, and the main branch also has a failure of Packer related testing. 
Is it freaky testing?

**Comment by yorinasub17** [comment]
@TsuyoshiUshio Yup it is flaky testing. The last two build failures contain tests that were mutually exclusive, so from a build perspective, this is good to go. Apologies that I haven't merged+released it yet. I need some time to formulate the release notes so I had been delaying doing that till later in the day. You should see this merged + released before EOD pacific time.

**Comment by TsuyoshiUshio** [comment]
@yorinasub17 Thank you!  
BTW, we are working on the subscription issue internally. Sorry for the late reply. :)
